clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -188,29 +193,27 @@ syscall_pt_trace_t::metadata_dump(pt_metadata_t metadata)
         return false;
     }
 
+    pt_metadata_buf_t to_write;
     /* Initialize the header of output buffer. */
-    syscall_pt_entry_t pdb_header[PT_METADATA_PDB_HEADER_ENTRY_NUM];
-    pdb_header[PDB_HEADER_PID_IDX].pid.type = SYSCALL_PT_ENTRY_TYPE_PID;
-    pdb_header[PDB_HEADER_PID_IDX].pid.pid = dr_get_process_id_from_drcontext(drcontext_);
-    pdb_header[PDB_HEADER_TID_IDX].tid.type = SYSCALL_PT_ENTRY_TYPE_THREAD_ID;
-    pdb_header[PDB_HEADER_TID_IDX].tid.tid = dr_get_thread_id(drcontext_);
-    pdb_header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_metadata_boundary.data_size =
+    to_write.header[PDB_HEADER_PID_IDX].pid.type = SYSCALL_PT_ENTRY_TYPE_PID;
+    to_write.header[PDB_HEADER_PID_IDX].pid.pid =
+        dr_get_process_id_from_drcontext(drcontext_);
+    to_write.header[PDB_HEADER_TID_IDX].tid.type = SYSCALL_PT_ENTRY_TYPE_THREAD_ID;
+    to_write.header[PDB_HEADER_TID_IDX].tid.tid = dr_get_thread_id(drcontext_);
+    to_write.header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_metadata_boundary.data_size =
         sizeof(pt_metadata_t);
-    pdb_header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_metadata_boundary.type =
+    to_write.header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_metadata_boundary.type =
         SYSCALL_PT_ENTRY_TYPE_PT_METADATA_BOUNDARY;
 
-    /* Write the buffer header to the output file */
-    if (write_file_func_(output_file_, pdb_header, PT_METADATA_PDB_HEADER_SIZE) == 0) {
+    /* We want to write the data in a single call to write_file_func_ for easier 3P
+     * integration.
```

Comment:
Please elaborate on why this makes integration easier: the write call is already indirected, so why would one vs 2 calls affect integration?  It should only affect overhead -- and there it is not clear which is better if the combining uses a copy and there is buffering already on the indirected write mechanism.

Reply:
Based on the implementation of the user-provided write func, it may not be acceptable to separate the header and data away from each other.

Reply:
Added to comment.

Reply:
Wait... I think it should be possible to still make the 3p implementation work with the multiple writes; though it gets a bit more complicated. Let me try and update this thread.

Reply:
I refactored 3p code to work with separate writes for header and data.

This allowed me to remove the heavy std::memcpy here. PTAL.